### DISCUSSION TITLE
License denied

### DISCUSSION THREAD

**Discussion by gonzus** [audiamus/BookLibConnect#65] [General]
[edit: added screenshot]

Hello, and thank you for this great application. I could connect with my Audible account and download / convert some of my books. However, one of them shows up with a key having a red strike on top, and the state says "License denied", even though the book was purchased like all my other books that do work. Here is a screenshot; any suggestions?

![audible](https://user-images.githubusercontent.com/208546/182167820-e1968797-e485-4950-a41c-25f1c13f5788.png)

Cheers!

**Comment by audiamus** [comment]
I would need the log file for this.

**Reply by gonzus** [reply]
I found these relevant lines. Please let me know if you need the whole file.

```
16:29:19.008 0001 [DownloadDecryptJob<ConversionContext>.getLicenseAndDownloadAsync] >>> Id=2, "The Count of Monte Cristo": license_denied                                                                 
16:29:19.080 0001 [AudibleApi.GetDownloadLicenseAndSaveAsync] >>> Id=2, "The Count of Monte Cristo": license_denied                                                                                        
16:29:19.081 0001 [AudibleApi.GetDownloadLicenseAndSaveAsync] Id=2, "The Count of Monte Cristo": license_denied; desired quality: High                                                                     
16:29:19.083 0001 [AudibleApi.GetDownloadLicenseAsync] >>> asin=B005GFRMBE                                                                                                                                 
16:29:19.485 0001 [AudibleApi.GetDownloadLicenseAsync] asin=B005GFRMBE, file="LicenseResponse_B005GFRMBE-2022_08_01-16_29_19.json"                                                                         
16:29:19.523 0001 [AudibleApi.GetDownloadLicenseAsync] <<< asin=B005GFRMBE                                                                                                                                 
16:29:19.525 0001 [AudibleApi.GetDownloadLicenseAndSaveAsync] Id=2, "The Count of Monte Cristo": license_denied; ArgumentNullException: "Value cannot be null. (Parameter 's')"                            
16:29:19.525 0001 [AudibleApi.GetDownloadLicenseAndSaveAsync] <<< Id=2, "The Count of Monte Cristo": license_denied                                                                                        
16:29:19.655 0001 [DownloadDecryptJob<ConversionContext>.getLicenseAndDownloadAsync] <<< Id=2, "The Count of Monte Cristo": license_denied                                                                 
16:29:19.657 0001 [DownloadDecryptJob<ConversionContext>.DownloadDecryptAndConvertAsync] <<< #conv=1
```

**Comment by audiamus** [comment]
The license json response may reveal more info. Can you upload `LicenseResponse_B005GFRMBE-2022_08_01-16_29_19.json`? (The encrypted key within - if there is one - is of no use to anyone but you.)

**Reply by gonzus** [reply]
Here you go, hope this helps. I had to append a `.txt` to the name, otherwise github complained, but the file is valid json.

[LicenseResponse_B005GFRMBE-2022_08_01-16_29_19.json.txt](https://github.com/audiamus/BookLibConnect/files/9235761/LicenseResponse_B005GFRMBE-2022_08_01-16_29_19.json.txt)

**Comment by audiamus** [comment]
Well, apparently the Book Lib Connect summarizing state looks correct in this case. License has not been granted, according to that response. The response has quite a few messages why the license was denied, like 
`Asin [B005GFRMBE] is not eligible for AYCL` (whatever AYCL may mean)
or 
`Customer [A2E9NDFXPILK1P] is not part of any plans.`
Is there anything specific about this title and your purchases/subscription/plan?

**Reply by gonzus** [reply]
I just checked; the Audible website shows for that book a button with `Unavailable` instead of `Listen now`, which is what appears for other titles. If I click on that button, it says `Check that your Audible membership isn’t cancelled or paused. If it isn't, this title may not be available anymore`.

I guess this means this is not a problem with Book Lib Connect, but some policy / permission / expiration in Audible (or in my account), which I will try to investigate. Sorry for the false alarm, and thanks for your quick responses!

**Reply by Sailfish** [reply]
It may be a regional issue. I'm from the U.S. and that title shows to be available for purchase. However, I have encountered a situation in the past when a previous title was no longer available since the author chose to remove it from Audible. Could it be that the title was pulled from your region by Audible, for reasons unknown?

From the log file, it appears that your title's ID is unavailable, see:
https://www.audible.com/pd/The-Count-of-Monte-Cristo-Audiobook/B005GFRMBE

whereas, the title ID Audible displays for the title and is available is:
https://www.audible.com/pd/The-Count-of-Monte-Cristo-Audiobook/B005GFQ5WQ

It could also be that author/Audible revised the title in some way and gave it a new ID, maybe?

**Reply by audiamus** [reply]
Well, Alexandre Dumas' "The Count of Monte Christo" is a classic from the 19th century. Author's rights have expired long since. The publisher is Naxos. They have quite a good reputation for publishing older literature as audio (and classical music, of course). They probably hold the production rights for this particular recording and perhaps fell out of love with Audible over this production and withdrew it.

**Reply by jschoen42** [reply]
AYCL are the **free** audiobooks from Audible (All You Can Listen) - not the purchased audiobooks

https://www.audible.com/about/newsroom/all-you-can-listen-membership-option-audible-plus-rolls-out-in-preview

for AYCL you need an active (paid) account for download, for the normal paid audiobooks not

**Reply by Sailfish** [reply]
Note: audiamus, yes, RightsOwner is a better term to use than author.

gonzus, if you've not recently downloaded that title (and the title had been updated) then the updated title should be available for you to re-download it. If you can, that should update your Audible title library, perhaps?

good luck

**Reply by gonzus** [reply]
I think what happened here is that this book was temporarily made available for download with a normal Audible subscription (instead of having to pay for the book), and now it was withdrawn.

This leads me to another question. I did download this book into my phone, and I can see the title there:
```
$ ls -l *MBE*
-rw-r--r--  1 gonzo  staff    41B Jul 31 22:26 B005GFRMBE.companion
-rw-r--r--  1 gonzo  staff   1.4G Jul 31 22:27 B005GFRMBE_ep6.aaxc
```
I would guess this file is encrypted, and now I am wondering: can I somehow process this AAXC file with Book Lib Connect? Or perhaps I could extract from the Book Lib Connect installation the activation code and manually enter that into AAX Audio Converter? If this is possible, how would I get the activation code from my already working installation of Book Lib Connect?

**Reply by audiamus** [reply]
> I would guess this file is encrypted, and now I am wondering: can I somehow process this AAXC file with Book Lib Connect? Or perhaps I could extract from the Book Lib Connect installation the activation code and manually enter that into AAX Audio Converter?

In theory, it's possible, but in practice, it's rather difficult.

The way aaxc files work, slightly simplified:

- When you register your device with Audible, i.e. your phone or a pseudo device like Book Lib Connect, Audible will give you a token. This token is specific to your device and your Amazon account. A different device, even under the same Amazon/Audible account, will result in a different token.
- That token will be used for all Audible API communication, to digitally sign every API request, as a means of authentication.
- When you request a license for a particular book and if that license is granted, Audible sends you a download link for an aaxc file and also a voucher. The voucher will hold the key for the aaxc file. The key will be specific to book, account and device.
- The voucher is encrypted. The device token is the the key to decrypt that voucher and obtain the aaxc key.
- The aaxc key will then be used to decrypt the aaxc file. 

The open source community has figured out this process a couple of years ago and published it on GitHub and other places. For Book Lib Connect I simply followed those specs and implemented the necessary processing steps. The last step, aaxc decryption, is carried out by [AAXClean](https://github.com/Mbucari/AAXClean), a C# port from the relevant section of FFmpeg.

So, for decrypting an aaxc file downloaded to your phone, you would need the token Audible issued for that phone and the voucher for that particular aaxc file in addition to the aaxc file itself. (I have read somewhere that one of the Chinese payware tools claims to be able to handle this, but I wouldn't trust any of those.)

Other than with the older aax format, the cipher key for aaxc is a fairly long one, additionally protected with a seed, and can be regarded as very secure. Those methods floating around the Web for guessing/enumerating an aax key from the file itself will all fail miserably with aaxc. 

By the way, Audible encryption is always common [AES](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard), for aaxc, aax or the voucher. Nothing proprietary there. Most CPUs, Intel/AMD or ARM, offer hardware acceleration for AES decryption. That makes it so fast.

**Reply by gonzus** [reply]
Ah, ok. I will just listen to that book on the phone for now.

Thanks again for all the information shared in this discussion!

**Comment by jfburrows** [comment]
Hi, fantastic program, huge thank you!! I managed to download about half of my library but the rest had the locked key I have had Audile for 15 years so it is more than 75 books, is there another way or downloading them>

**Reply by jfburrows** [reply]
[LicenseResponse_B002V8L7KQ-2022_10_22-22_23_24.txt](https://github.com/audiamus/BookLibConnect/files/9845674/LicenseResponse_B002V8L7KQ-2022_10_22-22_23_24.txt)

**Reply by audiamus** [reply]
Are those books with the license denied reply regular purchases? 

This is the relevant section of the json response:
```
        "license_denial_reasons": [
            {
                "message": "Asin [B002V8L7KQ] is not eligible for AYCL",
                "rejectionReason": "ContentEligibility",
                "validationType": "AYCL"
            },
            {
                "message": "Ownership: User [A1CDQ9ZI81S16D] has [Purchase] rights.",
                "validationType": "Ownership"
            },
            {
                "message": "Since ClientId is null, skipping client rights validation.AAAClientId[ApolloEnv:AudibleApiExternalRouterService/NA/ACLS/Prod], does not has access to asin[B002V8L7KQ].",
                "rejectionReason": "RequesterEligibility",
                "validationType": "Client"
            },
            {
                "message": "Customer [A1CDQ9ZI81S16D] is not part of any plans.",
                "rejectionReason": "RequesterEligibility",
                "validationType": "Membership"
            }
        ],
        "message": "License not granted to customer [A1CDQ9ZI81S16D] for asin [B002V8L7KQ]",
```

**Reply by jfburrows** [reply]
HI yes they were all purchased using the credits that I had in my account.

**Reply by audiamus** [reply]
If I understand the rejection message correctly, it says they cannot find a match between the customer ID and the book ID (ASIN).  

So, if, for instance, you returned a book, or the book is shared in a family plan and a family plan member returned the book, it will trigger such a rejection message. That's what has been reported here and for some of the other Audible client projects. 

Can you download the book with the Android or IOS app?

**Comment by 4NowInBerlin** [comment]
Hi, Thank you for the good work !
Since Audible pulled their Windows player, I have been left unable to download the books I purchased.
BookLibConnect has been a light in the dark !

I am in the process of rescuing my library and downloading every book with BookLibConnect. I have noticed that the books that I bought some years ago, which are now available as part of the "all you can listen" package, cannot be downloaded because of "license denied".
Same with the many children's books I bought with the 2-for-1 credit offer : "license denied".
Is there anything I can do ?
I should add that I am not very computer literate.
Thank you ! E.

**Reply by audiamus** [reply]
The log file may give a hint, and the json response file. It could be similar to https://github.com/audiamus/BookLibConnect/discussions/65#discussioncomment-3942186.

**Reply by 4NowInBerlin** [reply]
thank you !

**Comment by ozsmac** [comment]
UPDATE: I tried to re-add the book a few times in BookLibConnect and it ended up working.

Original message:
I've just had this occur for the first time.  Purchased a book about 40 minutes ago with a credit (no special deals), was able to download to my iPhone and play it.

My preference is to use Prologue client from my Plex setup, hence I zoomed to the amazing BookLibConnect to help me to use the audio book the way I wanted, however getting the red padlock also.

**Reply by dwellna** [reply]
Hi, if anyone is interested:
I just got the same error, "License denied" but the issue was that I had enabled multi part download which is no longer supported for recently released books on Audible.
I saw this exception in the log file:
"HttpRequestException: "Response status code does not indicate success: 404 (Not Found)."
{"message":"Audio Part asins are no longer supported."}"

So disabling "Multi Part Download" fixed the issue.

**Reply by ksmith9226** [reply]
@dwellna 
 
 Thank you, 
 Once I unchecked 'Multi Part Download', my license issue was fixed.

**Reply by festivalist** [reply]
I have to thank you as well! That did it!

**Reply by lukasantares** [reply]
Perfect, disabling “Multipart Download” seems to fix the problem!
If I had found this post earlier, I could have saved myself two hours of unsuccessful troubleshooting by simply checking a box ;)
